From 9bc303348279c9ba400da01b2eec44fdb19e0fd0 Mon Sep 17 00:00:00 2001
From: junyulai <junyulai@google.com>
Date: Mon, 25 Jun 2018 21:15:58 +0800
Subject: [PATCH 08/39] Data activity didn't reflect UDP traffic

Currently, the DataActivity will use the same source with data
stall detection to count packet traffic, which is TCP only.
However, more and more modern applications uses UDP for
communication such as QUIC and streaming. And the data icon
arrows did not reflect the truth that packet is transmitting.
It should be more accurate that data icon arrows should count
all packet traffic.

Note that data stall detection remains the same for now since
we don't want to break functionality for below cases:
1. Out-going UDP packets trigger false positive, such as live
   broadcast apps or b/7903145
2. Incoming local communications cause false negative, such as
   IPv6 Neighbor Discovery messages between modem & AP.

Bug: 110773754
Test: 1. play youtube and see data icon arrows in QuickSetting
      2. enable updateDataActivity debug log and make sure it
         counts.
      2. runtest frameworks-net

Change-Id: Id2f763f396b329ab82af09bb5e084af7201f9d7c
---
 .../telephony/dataconnection/DcTracker.java     | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/src/java/com/android/internal/telephony/dataconnection/DcTracker.java b/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
index 38d3852b1..1f4ada6c4 100644
--- a/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
+++ b/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
@@ -356,10 +356,21 @@ public class DcTracker extends Handler {
             return "{txSum=" + txPkts + " rxSum=" + rxPkts + "}";
         }
 
-        public void updateTxRxSum() {
+        /**
+         * Get Tcp Tx/Rx packet count from TrafficStats
+         */
+        public void updateTcpTxRxSum() {
             this.txPkts = TrafficStats.getMobileTcpTxPackets();
             this.rxPkts = TrafficStats.getMobileTcpRxPackets();
         }
+
+        /**
+         * Get total Tx/Rx packet count from TrafficStats
+         */
+        public void updateTotalTxRxSum() {
+            this.txPkts = TrafficStats.getMobileTxPackets();
+            this.rxPkts = TrafficStats.getMobileRxPackets();
+        }
     }
 
     private void onActionIntentReconnectAlarm(Intent intent) {
@@ -4508,7 +4519,7 @@ public class DcTracker extends Handler {
 
         TxRxSum preTxRxSum = new TxRxSum(mTxPkts, mRxPkts);
         TxRxSum curTxRxSum = new TxRxSum();
-        curTxRxSum.updateTxRxSum();
+        curTxRxSum.updateTotalTxRxSum();
         mTxPkts = curTxRxSum.txPkts;
         mRxPkts = curTxRxSum.rxPkts;
 
@@ -4681,7 +4692,7 @@ public class DcTracker extends Handler {
         long sent, received;
 
         TxRxSum preTxRxSum = new TxRxSum(mDataStallTxRxSum);
-        mDataStallTxRxSum.updateTxRxSum();
+        mDataStallTxRxSum.updateTcpTxRxSum();
 
         if (VDBG_STALL) {
             log("updateDataStallInfo: mDataStallTxRxSum=" + mDataStallTxRxSum +
-- 
2.17.1

